### ISSUE TITLE
Only one GPU is being used on SLURM cluster

### ISSUE THREAD

**Issue by vandrw** [Lightning-AI/lightning#18950]
### Bug description

While trying to train a HuggingFace model on a SLURM cluster, I am facing some issues with running on multiple GPUs. I have tried both FSDP and DeepSpeed, and while Lightning can identify that there are multiple GPUs (`CUDA_VISIBLE_DEVICES: [0,1,2,3]`), it doesn't use them. To ensure that it is not my code causing the problem, I have ran some tests with an example script from the Lightning documentation, resulting in the same outcome.

Before this, I was able to get the script to use 2/4 GPUs, or even 3/4, but it seems to be random. I am not sure how to debug further, but I would be happy to help if you can provide any feedback.

### What version are you seeing the problem on?

v2.1

### How to reproduce the bug
I used the following Python script:
```python
Using a default example from the docs:

import torch
import torch.nn as nn
import torch.nn.functional as F
from torch.utils.data import DataLoader

import lightning as L
from lightning.pytorch.strategies import FSDPStrategy
from lightning.pytorch.demos import Transformer, WikiText2


# Fast: Delays the model creation until Trainer can place it on GPU
class LanguageModel(L.LightningModule):
    def __init__(self, vocab_size):
        super().__init__()
        self.vocab_size = vocab_size
        self.model = None

    def configure_model(self):
        if self.model is not None:
            return
        self.model = Transformer(  # 1B parameters
            vocab_size=self.vocab_size,
            nlayers=32,
            nhid=4096,
            ninp=1024,
            nhead=64,
        )

    def training_step(self, batch):
        input, target = batch
        output = self.model(input, target)
        loss = F.nll_loss(output, target.view(-1))
        self.log("train_loss", loss, prog_bar=True)
        return loss

    def configure_optimizers(self):
        return torch.optim.Adam(self.parameters(), lr=0.1)


L.seed_everything(42)

# Data
dataset = WikiText2()
train_dataloader = DataLoader(dataset)

# Model
model = LanguageModel(vocab_size=dataset.vocab_size)

# Trainer
trainer = L.Trainer(accelerator="cuda", devices=4, strategy=FSDPStrategy())
trainer.fit(model, train_dataloader)
trainer.print(torch.cuda.memory_summary())
```

This is the bash script I am using to schedule the job:
```bash
#!/bin/bash
#SBATCH --job-name=test_parallel
#SBATCH --output=parallel_test_%j.log
#SBATCH --partition=gpu
#SBATCH --nodes=1
#SBATCH --gpus=4
#SBATCH --cpus-per-task=64
#SBATCH --mem=256G
#SBATCH --signal=SIGUSR1@90
#SBATCH --time=0-01:00:00

module load 2022 Python/3.10.4-GCCcore-11.3.0

source $HOME/linguamol/.venv/bin/activate

python collect_env_details.py

# I tried both. None change the outcome of GPU usage.
# srun --nodes=1 --gpus=4 --cpus-per-task=64 python demo_parallelism.py
srun python demo_parallelism.py

deactivate
```

### Error messages and logs

Python script log:
```
[rank: 0] Seed set to 42
GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
HPU available: False, using: 0 HPUs
/gpfs/home/user/project/.venv/lib/python3.10/site-packages/lightning/pytorch/trainer/connectors/logger_connector/logger_connector.py:67: Starting from v1.9.0, `tensorboardX` has been removed as a dependency of the `lightning.pytorch` package, due to potential conflicts with other packages in the ML ecosystem. For this reason, `logger=True` will use `CSVLogger` as the default logger, unless the `tensorboard` or `tensorboardX` packages are found. Please `pip install lightning[extra]` or one of them to enable TensorBoard support by default
/gpfs/home/user/project/.venv/lib/python3.10/site-packages/lightning/pytorch/loops/utilities.py:72: `max_epochs` was not set. Setting it to 1000 epochs. To train without an epoch limit, set `max_epochs=-1`.
[rank: 0] Seed set to 42
Initializing distributed: GLOBAL_RANK: 0, MEMBER: 1/1
----------------------------------------------------------------------------------------------------
distributed_backend=nccl
All distributed processes registered. Starting with 1 processes
----------------------------------------------------------------------------------------------------

You are using a CUDA device ('NVIDIA A100-SXM4-40GB') that has Tensor Cores. To properly utilize them, you should set `torch.set_float32_matmul_precision('medium' | 'high')` which will trade-off precision for performance. For more details, read https://pytorch.org/docs/stable/generated/torch.set_float32_matmul_precision.html#torch.set_float32_matmul_precision
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0,1,2,3]

  | Name  | Type        | Params
--------------------------------------
0 | model | Transformer | 1.0 B 
--------------------------------------
1.0 B     Trainable params
5.1 M     Non-trainable params
1.0 B     Total params
4,055.534 Total estimated model params size (MB)
SLURM auto-requeueing enabled. Setting signal handlers.
/gpfs/home/user/project/.venv/lib/python3.10/site-packages/lightning/pytorch/trainer/connectors/data_connector.py:441: The 'train_dataloader' does not have many workers which may be a bottleneck. Consider increasing the value of the `num_workers` argument` to `num_workers=15` in the `DataLoader` to improve performance.

Training: |          | 0/? [00:00<?, ?it/s]
Training:   0%|          | 0/14919 [00:00<?, ?it/s]
...
```

Output of `nvidia-smi`:
```
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.104.12             Driver Version: 535.104.12   CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  NVIDIA A100-SXM4-40GB          On  | 00000000:31:00.0 Off |                  Off |
| N/A   36C    P0             224W / 400W |  20672MiB / 40960MiB |     47%      Default |
|                                         |                      |             Disabled |
+-----------------------------------------+----------------------+----------------------+
|   1  NVIDIA A100-SXM4-40GB          On  | 00000000:32:00.0 Off |                  Off |
| N/A   29C    P0              48W / 400W |      8MiB / 40960MiB |      0%      Default |
|                                         |                      |             Disabled |
+-----------------------------------------+----------------------+----------------------+
|   2  NVIDIA A100-SXM4-40GB          On  | 00000000:CA:00.0 Off |                  Off |
| N/A   29C    P0              52W / 400W |      8MiB / 40960MiB |      0%      Default |
|                                         |                      |             Disabled |
+-----------------------------------------+----------------------+----------------------+
|   3  NVIDIA A100-SXM4-40GB          On  | 00000000:E3:00.0 Off |                  Off |
| N/A   29C    P0              50W / 400W |      8MiB / 40960MiB |      0%      Default |
|                                         |                      |             Disabled |
+-----------------------------------------+----------------------+----------------------+
                                                                                         
+---------------------------------------------------------------------------------------+
| Processes:                                                                            |
|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
|        ID   ID                                                             Usage      |
|=======================================================================================|
|    0   N/A  N/A   1051848      C   .../user/project/.venv/bin/python       20658MiB   |
+---------------------------------------------------------------------------------------+
```

### Environment

<details>
  <summary>Current environment</summary>

* CUDA:
	- GPU:
		- NVIDIA A100-SXM4-40GB
		- NVIDIA A100-SXM4-40GB
		- NVIDIA A100-SXM4-40GB
		- NVIDIA A100-SXM4-40GB
	- available:         True
	- version:           12.1
* Lightning:
	- lightning:         2.1.0
	- lightning-utilities: 0.9.0
	- pytorch-lightning: 2.1.0
	- pytorch-triton:    2.1.0+6e4932cda8
	- torch:             2.2.0.dev20231030+cu121
	- torchmetrics:      1.2.0
* Packages:
	- aiohttp:           3.8.6
	- aiosignal:         1.3.1
	- antlr4-python3-runtime: 4.9.3
	- appdirs:           1.4.4
	- async-timeout:     4.0.3
	- attrs:             23.1.0
	- certifi:           2023.7.22
	- charset-normalizer: 3.3.0
	- click:             8.1.7
	- deepspeed:         0.11.2+f0604078
	- diffusers:         0.21.4
	- docker-pycreds:    0.4.0
	- fast-tokenizer-sf: 0.1.0
	- filelock:          3.13.1
	- frozenlist:        1.4.0
	- fsspec:            2023.10.0
	- gitdb:             4.0.10
	- gitpython:         3.1.37
	- hjson:             3.1.0
	- huggingface-hub:   0.17.3
	- idna:              3.4
	- importlib-metadata: 6.8.0
	- jinja2:            3.1.2
	- lightning:         2.1.0
	- lightning-utilities: 0.9.0
	- linguamol:         0.0.1
	- markupsafe:        2.1.3
	- maturin:           1.3.0
	- mpmath:            1.3.0
	- multidict:         6.0.4
	- networkx:          3.2.1
	- ninja:             1.11.1.1
	- numpy:             1.26.1
	- nvidia-cublas-cu12: 12.1.3.1
	- nvidia-cuda-cupti-cu12: 12.1.105
	- nvidia-cuda-nvrtc-cu12: 12.1.105
	- nvidia-cuda-runtime-cu12: 12.1.105
	- nvidia-cudnn-cu12: 8.9.2.26
	- nvidia-cufft-cu12: 11.0.2.54
	- nvidia-curand-cu12: 10.3.2.106
	- nvidia-cusolver-cu12: 11.4.5.107
	- nvidia-cusparse-cu12: 12.1.0.106
	- nvidia-nccl-cu12:  2.18.1
	- nvidia-nvjitlink-cu12: 12.3.52
	- nvidia-nvtx-cu12:  12.1.105
	- omegaconf:         2.3.0
	- packaging:         23.2
	- pathtools:         0.1.2
	- pillow:            10.1.0
	- pip:               23.3.1
	- polars:            0.19.8
	- protobuf:          4.24.4
	- psutil:            5.9.6
	- py-cpuinfo:        9.0.0
	- pydantic:          1.10.13
	- pynvml:            11.5.0
	- pytorch-lightning: 2.1.0
	- pytorch-triton:    2.1.0+6e4932cda8
	- pyyaml:            6.0.1
	- rdkit:             2023.3.3
	- regex:             2023.10.3
	- requests:          2.31.0
	- safetensors:       0.4.0
	- selfies:           2.1.1
	- sentry-sdk:        1.32.0
	- setproctitle:      1.3.3
	- setuptools:        58.1.0
	- six:               1.16.0
	- smmap:             5.0.1
	- sympy:             1.12
	- tokenizers:        0.14.1
	- tomli:             2.0.1
	- torch:             2.2.0.dev20231030+cu121
	- torchmetrics:      1.2.0
	- tqdm:              4.66.1
	- transformers:      4.34.0
	- triton:            2.1.0
	- typing-extensions: 4.8.0
	- urllib3:           2.0.6
	- wandb:             0.15.12
	- yarl:              1.9.2
	- zipp:              3.17.0
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- ELF
	- processor:         x86_64
	- python:            3.10.4
	- release:           4.18.0-372.70.1.el8_6.x86_64
	- version:           #1 SMP Thu Aug 10 14:05:06 EDT 2023

</details>

cc @awaelchli

**Comment by awaelchli** [comment]
@vandrw Make sure to closely follow the instructions here: 
https://lightning.ai/docs/pytorch/stable/clouds/cluster_advanced.html

There it shows the important variables you need to set, in particular:

```
#SBATCH --nodes=4             # This needs to match Trainer(num_nodes=...)
#SBATCH --ntasks-per-node=8   # This needs to match Trainer(devices=...)
```

Which you haven't done. Can you check please?

**Comment by vandrw** [comment]
Oh, that is embarrassing. You are right! Thanks a lot for pointing it out.

Regarding the cpus-per-task, can you help me understand how the DataLoaders are being used in this parallel scenario? I assume that only one will be instantiated, and each task will use the provided CPUs to access it and serve data to the GPU?

If so, in your experience, how many CPUs per task would be enough for training while using A100-40GB GPUs?

**Comment by awaelchli** [comment]
There will be N main processes (or "tasks" in the slurm terminology) per machine if you select ntasks-per-node=N. If you then ask the SLURM scheduler to give you cpus-per-task=64 CPUs per task, then that's a lot of CPUs per task. Are you sure you have this many? If not required, I would just remove that requirement and just take the whole machine :) 

How many CPU cores you want to use for loading your data can be specified in the PyTorch dataloader `DataLoader(num_workers=...)` and the value needs to be tuned to your needs.

**Comment by julius-thomas** [comment]
I am facing a similar issue here.
I am training on an off-perm slurm cluster, and want to use a single node with 4 GPUs.
After reading the [documentation](https://lightning.ai/docs/pytorch/stable/clouds/cluster_advanced.html) I am slightly confused. After setting `--ntasks-per-node=4` equal to `Trainer(devices=4, ...),` I am now spawning 4 main processes that each request 4 GPUs which naturally fails.  Ideally, I would like a single task with 4 GPUs as would be the case on a local machine.

**Comment by vandrw** [comment]
Hey @julius-thomas! In SLURM, there are different ways to request GPUs. For example, you can request using the `--gpus=4` flag, `--gpus-per-node=4`, or `--gpus-per-task=4`. In my case, I set it up like this: 
```
#SBATCH --partition=gpu
#SBATCH --nodes=1
#SBATCH --gpus-per-node=4 (this could also include the type of GPU, such as a100:4)
#SBATCH --ntasks-per-node=4
```

Now, that would give you 1 node with 4 GPUs. To make sure that this is used properly by the Trainer, you need to also provide the number of nodes:
```
trainer = L.Trainer(
        accelerator="auto",
        strategy="auto",
        devices=4,
        num_nodes=1,
    )
```

I hope this helps!

**Comment by julius-thomas** [comment]
Sadly not. This is my batch script.
```
#SBATCH --nodes=1
#SBATCH --ntasks-per-node=4
#SBATCH --gres=gpu:4
#SBATCH --hint=nomultithread
```
and this is my trainer:
```
    trainer = Trainer(
        devices=4,
        strategy='ddp'
    )
```

The error thrown in this configuration is:
```
raise MisconfigurationException(
lightning_fabric.utilities.exceptions.MisconfigurationException: You requested gpu: [0, 1, 2, 3]
```
4 Times, once for each task.

**Comment by vandrw** [comment]
@julius-thomas 
I just realized that the default value for num_nodes is 1, so my previous statement shouldn't make a difference. It does strike me as off that it only sees [0], [1], ... on each task instead of showing [0, 1, 2, 3]. Can you try running the example transformer I provided above, with the following script? You will have to change the Python version and the virtual environment path to yours:

```
#!/bin/bash
#SBATCH --job-name=test_parallelism
#SBATCH --partition=gpu
#SBATCH --nodes=1
#SBATCH --gpus=4
#SBATCH --ntasks-per-node=4
#SBATCH --mem=128G
#SBATCH --time=00:30:00

module purge
module load Python/3.10.8-GCCcore-12.2.0
source $HOME/project/.venv/bin/activate

srun python demo_parallelism.py

deactivate
```
